Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+define([
+    './appendForwardSlash',
+    './Check',
+    './clone',
+    './combine',
+    './defaultValue',
+    './defined',
+    './defineProperties',
+    './freezeObject',
+    './getAbsoluteUri',
+    './getBaseUri',
+    './getExtensionFromUri',
+    './isDataUri',
+    './objectToQuery',
+    './queryToObject',
+    '../ThirdParty/Uri',
+    '../ThirdParty/when'
+], function(appendForwardSlash,
+            Check,
+            clone,
+            combine,
+            defaultValue,
+            defined,
+            defineProperties,
+            freezeObject,
+            getAbsoluteUri,
+            getBaseUri,
+            getExtensionFromUri,
+            isDataUri,
+            objectToQuery,
+            queryToObject,
+            Uri,
+            when) {
+    'use strict';
+
+    /**
+     * @private
+     */
+    function parseQuery(queryString) {
+        if (queryString.length === 0) {
+            return {};
+        }
+
+        // Special case we run into where the querystring is just a string, not key/value pairs
+        if (queryString.indexOf('=') === -1) {
+            var result = {};
+            result[queryString] = undefined;
+            return result;
+        }
+
+        return queryToObject(queryString);
+    }
+
+    /**
+     * @private
+     */
+    function stringifyQuery(queryObject) {
+        var keys = Object.keys(queryObject);
+
+        // We have 1 key with an undefined value, so this is just a string, not key/value pairs
+        if (keys.length === 1 && !defined(queryObject[keys[0]])) {
+            return keys[0];
+        }
+
+        return objectToQuery(queryObject);
+    }
+
+    /**
+     * @private
+     */
+    function encodeValues(object) {
+        var result = {};
+        for(var key in object) {
+            if (object.hasOwnProperty(key)) {
+                result[key] = encodeURIComponent(object[key]);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * A resource the location and any other parameters we need to retrieve it or create derived resources.
+     *
+     * @param {Object} options An object with the following properties
+     * @param {String} options.url The url of the resource.
+     * @param {Object} [options.queryParameters] An object containing query parameters that will be sent when retrieving the resource.
+     * @param {Object} [options.templateValues] Key/Value pairs that are used to replace template values (eg. {x}).
+     * @param {Object} [options.headers={}] Additional HTTP headers that will be sent.
+     * @param {String} [options.responseType] The type of response.
+     * @param {String} [options.method='GET'] The method to use.
+     * @param {Object} [options.data] Data that is sent with the resource request.
```

Comment:
Perhaps specifically mentioned when `data` is used, I assume it's just for `PUT`/`POST`?